File: LeetCode Solutions/LeetCoding Challenge/May 2020 LeetCoding Challenge/Range Sum Query 2D - Immutable.py
class NumMatrix:

    def __init__(self, matrix: List[List[int]]):
        ylength, xlength = len(matrix) + 1, len(matrix[0]) + 1
        self.arr = [[0] * xlength for _ in range(ylength)]
        for i in range(1, ylength):
            for j in range(1, xlength):
                self.arr[i][j] = matrix[i-1][j-1] + self.arr[i-1][j] + self.arr[i][j-1] - self.arr[i-1][j-1]

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        return self.arr[row2+1][col2+1] - self.arr[row2+1][col1] - self.arr[row1][col2+1] + self.arr[row1][col1]


# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# param_1 = obj.sumRegion(row1,col1,row2,col2)

File: LeetCode Solutions/Study Plan/Programming Skills/Programming Skills II/Range Sum Query 2D - Immutable.java
class NumMatrix {
    long[][] arr;
    public NumMatrix(int[][] matrix) {
        int ylength = matrix.length + 1, xlength = matrix[0].length + 1;
        arr = new long[ylength][xlength];
        for (int i = 1; i < ylength; i++)
            for (int j = 1; j < xlength; j++)
                arr[i][j] = matrix[i-1][j-1] + arr[i-1][j] + arr[i][j-1] - arr[i-1][j-1];
    }
    
    public int sumRegion(int row1, int col1, int row2, int col2) {
        return (int)(arr[row2+1][col2+1] - arr[row2+1][col1] - arr[row1][col2+1] + arr[row1][col1]);
    }
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix obj = new NumMatrix(matrix);
 * int param_1 = obj.sumRegion(row1,col1,row2,col2);
 */